src/sim/TrackInitData.hh

```diff
@@ -254,12 +255,11 @@ void resize(
     // Allocate device data
     auto capacity = params.capacity;
     make_builder(&data->initializers.storage).resize(capacity);
```

Comment:
Would it make to call `CollectionBuilder::reserve` here? (if not, should be documented?)

Reply:
Most of the `CollectionBuilder` methods (all except `resize`) only work on host data because the underlying storage for a `Collection` on host is a vector but on device is a `DeviceVector`, which doesn't have all the same functionality (`reserve`, `insert`, `push_back`, etc.). `CollectionBuilder::resize` can be used to allocate (uninitialized) device memory without first allocating it on host. Like you said below, it doesn't behave like `vector::resize` since it always allocates new memory and doesn't preserve any data that might have been there, so the name is a little misleading.

The `initializers` here are also a special case since they are this (kind of awkward) type `ResizableData`, which is a `Collection` with a fixed capacity and dynamic size. So you're right, the line above Is really reserving space for the initializers (maybe encapsulating that in a `ResizableData::reserve` would make it a little clearer?) and line 262 is calling `ResizableData::resize`.

Reply:
>  it doesn't behave like vector::resize

Hence my recommendation that we consider renaming it :)

Reply:
Maybe @sethrj wants to chime in, since he named it ;)

Reply:
It's more like `clear_and_resize` than reserve. The method, as part of the "builder", was meant as part of the initial construction of state containers, resizing empty collections to hold values for the number of states. So in that situation it does act as resize. Since I think our device vectors assert that we use trivially copyable data, we *could* extend collections so that they have a legitimate `resize` capability: allocate-copy-deallocate. However, most of our containers don't need the general resize capability, and that carries the penalty of needing an extra size for every collection to track its capacity.